src/business_logic/state/update_tracker_state.rs

```diff
@@ -0,0 +1,139 @@
+use super::{
+    state_api::{State, StateReader},
+    state_api_objects::BlockInfo,
+    state_cache::StorageEntry,
+};
+
+// An implementation of the SyncState API that wraps another SyncState object and contains a cache.
+// All requests are delegated to the wrapped SyncState, and caches are maintained for storage reads
+// and writes.
+
+// The goal of this implementation is to allow more precise and fair computation of the number of
+// storage-writes a single transaction preforms for the purposes of transaction fee calculation.
+// That is, if a given transaction writes to the same storage address multiple times, this should
+// be counted as a single storage-write. Additionally, if a transaction writes a value to storage
+// which is equal to the initial value previously contained in that address, then no change needs
+// to be done and this should not count as a storage-write.
+pub(crate) struct UpdatesTrackerState<T: State> {
+    pub(crate) state: T,
+    pub(crate) storage_initial_values: HashMap<StorageEntry, u64>,
+    pub(crate) storage_writes: HashMap<StorageEntry, u64>,
+}
+
+impl<T: State + StateReader> UpdatesTrackerState<T> {
+    pub fn new(state: T) -> Self {
+        UpdatesTrackerState {
+            state,
+            storage_initial_values: HashMap::new(),
+            storage_writes: HashMap::new(),
+        }
+    }
+
+    // This method writes to a storage cell and updates the cache accordingly. If this is the first
+    // access to the cell (read or write), the method first reads the value at that cell and caches
+    // it.
+    // This read operation is necessary for fee calculation. Because if the transaction writes a
+    // value to storage that is identical to the value previously held at that address, then no
+    // change is made to that cell and it does not count as a storage-change in fee calculation.
+
+    pub fn set_storage_at(
+        &mut self,
+        contract_address: Felt,
+        key: u64,
+        value: Felt,
+    ) -> Result<(), StateError> {
+        let address_key_pair = (contract_address, key);
+        if !self.was_accessed(&address_key_pair) {
+            let val = self.state.get_storage_at(&address_key_pair)?;
+            let value = val
+                .to_u64()
+                .ok_or_else(|| StateError::ConversionError(value.clone()))?;
+            self.storage_initial_values
+                .insert(address_key_pair.clone(), value);
+        }
+
+        let stor_value = value
+            .to_u64()
+            .ok_or_else(|| StateError::ConversionError(value.clone()))?;
+        self.storage_writes
+            .insert(address_key_pair.clone(), stor_value);
+        self.state.set_storage_at(&address_key_pair, value);
+        Ok(())
+    }
+
+    pub fn get_storage_at(&mut self, contract_address: Felt, key: u64) -> Result<Felt, StateError> {
+        let address_key_pair = (contract_address.clone(), key);
+        let was_accessed = !self.was_accessed(&address_key_pair);
```

Comment:
Since we are negating the `was_accessed` we should use a different name, something like:
```suggestion
        let was_not_accessed = !self.was_accessed(&address_key_pair);
```

Reply:
fixed